### ISSUE TITLE
Custom META - JSON Schema support for meta in `exo spec`

### ISSUE THREAD

**Issue by willcharlton** [exosite/exoline#141]
In YAML spec files, we can specify JSON schema of dataports, but how about specifying the JSON schema of meta and/or subelements of meta (i.e. location)?

```
"meta": {
  "location": {
    "gps": {
      "long": "-93.28874739999999",
      "lat": "44.98014269999999"
    },
    "address": {
      "city": "Minneapolis",
      "name": "Exosite",
      "street": "275 Market Street",
      "zip": "55405",
      "state": "MN"
    }
  },
  "device": {
    "type": "vendor",
    "model": "my_model",
    "sn": "00:11:22:33:44:55",
    "vendor": "willcharlton"
  },
  "activetime": "5",
  "timezone": "America/New_York",
  "DeviceType": "generic",
  "DeviceTypeName": "generic",
  "DeviceTypeID": "0000000002",
  "Timezone": "America/New_York",
  "Location": "USA"
},
```

In this example, I have a JSON schema for the 'location' subelement in meta. I would like to specify that in the spec file. Is that even supported?

I'd like to know for a few reasons:

1. It would make an exo spec meta --update <path_to_spec> (or something to that effect) a worthwhile implementation because I've carved out code pathways for doing this manually, and it was really useful.
2. It is really good practice to document schemas like this in our spec files.
3. It would be good to have default behavior keep all of the 'required' meta (if there even is such a thing) intact, but allow the user/developer to define custom meta. This was useful on one project and is a decent poor-man's, long-term key-value store.

Thoughts?

Thanks,

~Will

**Comment by willcharlton** [comment]
To the point about race conditions that can exist when updating meta. Race conditions can exist pretty much anywhere I think, but, as it happens, meta is worse because all history is lost (if you use a dataport, you have the time-series data). It would be up to the developer to use with caution.

From a philosophical standpoint, some data makes more sense in meta because the nature of the data might be metadata about the physical device. It might be worth looking into locking meta or only allowing parent ciks the ability to update a client's meta (should a device be able to update its own meta?).

From an implementation standpoint, the code I developed code to update key:values that wasn't strictly an update to the meta. It first retrieved the latest meta and either added missing elements or updated them, then it would re-commit the updated meta. So, in theory, it would be really hard to screw up the existing meta - you'd only be adding or updating. The race-condition would still exist in the case of two processes trying to update meta at roughly the same time.

In a general sense though, meta shouldn't be updated that often. When a device moves location, if its serial number changes, other Earth-shattering changes a device can go through.

There's a part of me that wants to create a 'meta' dataport on all client models for the sake of a paper-trail, but that actually puts a cost burden on the device. I figure, we have meta, let's use it.

**Comment by dweaver** [comment]
Yeah. Race conditions are unlikely if you do info/modify/update as you suggest but potentially quite bad. The same issues apply to both meta and dataport.

meta can only be updated by the parent, so a device can't update its own meta. 

None of this is to say we shouldn't have JSONSchema for meta in spec. We use meta for things in multiple tools-- Exoline, Portals, Exo (the iOS app). The concurrent update issue is just something to be aware of when designing systems.

**Comment by tadpol** [comment]
To me, there are two separate things here: validation and updating.  The spec command really should be able to validate the contents of the meta.  I think really we're just waiting for someone with time to implement it.

**Comment by dweaver** [comment]
What @tadpol said. Good feature, someone with time or inclination should do it.
